packages/aws-cdk-lib/aws-stepfunctions/lib/states/custom-state.ts

```diff
@@ -74,11 +74,16 @@ export class CustomState extends State implements IChainable, INextable {
       ...this.renderRetryCatch(),
     };
 
-    // merge the Retry field defined in the stateJson into the state
+    // Retriers and Catchers can be specified directly in the stateJson or indirectly to the construct with addRetry() and addCatch().
+    // renderRetryCatch() only renders the indirectly supplied Retriers and Catchers, so we need to manually merge in those directly in the stateJson
     if (Array.isArray(this.stateJson.Retry)) {
       state.Retry = Array.isArray(state.Retry) ? [...state.Retry, ...this.stateJson.Retry] : [...this.stateJson.Retry];
     }
 
+    if (Array.isArray(this.stateJson.Catch)) {
+      state.Catch = Array.isArray(state.Catch) ? [...state.Catch, ...this.stateJson.Catch] : [...this.stateJson.Catch];
+    }
+
```

Comment:
Question I probably should've asked when reviewing https://github.com/aws/aws-cdk/pull/29403 but is there a reason we allow customers to define both and not just throw an error here saying they can only define one?

Reply:
Discussed offline, we decided the best way would be for addRetry and addCatch to have the same behavior but to add warnings so the customer understands how the merge works.